### ISSUE TITLE
Database does not return all rows on select query

### ISSUE THREAD

**Issue by danchupin** [cockroachdb/cockroach#74736]
**Describe the problem**

We find a very strange query behaviour with sql like `select * from table where id in (...140k ids)`  (all ids are exist in table). This query returns only part of data, about 28k of 140k in our case. But when we divide query on batches (10k ids per batch for example) and gather results, then we have all expected rows.
**Important note**: we almost sure that we started to experience such behaviour when we upgrade db from 21.1.7 to 21.2.3

[Thread in community slack for history.](https://cockroachdb.slack.com/archives/CP4D9LD5F/p1641991038091300)

**To Reproduce**

What did you do? Describe in your own words.

If possible, provide steps to reproduce the behavior:

1. Create table with schema provided below (it was obfuscated manually so I apologize for inaccuracies)
2. Populate it with some random data (can't provide real data)
3. Execute query `SELECT id FROM my_table WHERE id IN (100k+ ids)` and capture rows count
4. Execute the same query multiple times with small batches (e.g. 10k) and gather total rows count
4. See described behaviour

**Expected behavior**
Query should return all rows

**Additional data / screenshots**
<details>
<summary>Python script (just for better understanding the case)</summary>

```python
import psycopg2

filename = 'ids.txt' # file with 140k identifiers

with open(filename) as f:
    ids = []
    for line in f:
        ids.extend(line.split(','))

print(len(ids))  # 141648

# make sure all IDs are unique
assert len(ids) == len(set(ids))

conn = psycopg2.connect(
    "dbname=db user=root port=26257 host=host password=pass")

cur = conn.cursor()

# Execute a query
ids_list = ', '.join(ids)
cur.execute(f'SELECT id FROM my_table WHERE id ({ids_list})')
records = cur.fetchall()

print(len(records))  # 28049 :-(

# Divide the input on batches

records = []
batch_size = 10_000

while ids:
    ids_list = ', '.join(ids[:batch_size])
    cur.execute(f'SELECT id FROM my_table WHERE id IN ({ids_list})')
    records.extend(cur.fetchall())

    ids = ids[batch_size:]

print(len(records))  # 141648 :-)
```
</details>

<details>
<summary>Query plan</summary>

```
planning time: 877ms
execution time: 10.2s
distribution: local
vectorized: true
rows read from KV: 28,049 (0 B)
cumulative time spent in KV: 10.2s
maximum memory usage: 36 MiB
network usage: 0 B (0 messages)
regions: region1

• scan
  columns: (id int)
  nodes: n8
  regions: region1
  actual row count: 28,049
  vectorized batch count: 28
  KV time: 10.2s
  KV contention time: 0µs
  KV rows read: 28,049
  KV bytes read: 0 B
  estimated row count: 141,069 (0.06% of the table; stats collected 3 hours ago)
  table: my_table@primary
  spans: [very large number of spans]
```
</details>

<details>
<summary>Schema (obfuscated)</summary>

```sql
CREATE TABLE public.my_table (
	id INT8 NOT NULL DEFAULT nextval('public.id_seq'::REGCLASS),
	col1 INT8 NULL,
	col2 STRING NULL,
	col3 INT8 NULL,
	col4 STRING NULL,
	col5 INT8 NULL,
	col6 STRING NULL,
	col7 INT8 NULL,
	col8 STRING NULL,
	col9 INT8 NULL,
	col10 STRING NULL,
	col11 INT8 NULL,
	col12 STRING NULL,
	col13 BOOL NULL,
	col14 BOOL NULL,
	col15 INT8 NULL,
	col16 TIMESTAMPTZ NULL,
	col17 INT8 NULL,
	col18 STRING NULL,
	col19 STRING NULL,
	col20 BOOL NULL DEFAULT false,
	col21 BOOL NULL DEFAULT false,
	col22 INT2 NOT NULL DEFAULT 1:::INT8,
	col23 INT8 NULL DEFAULT 0:::INT8,
	col24 INT8 NULL,
	col25 INT8 NULL DEFAULT 0:::INT8,
	col26 INT8 NULL,
	col27 TIMESTAMPTZ NULL,
	col28 TIMESTAMPTZ NULL,
	col29 TIMESTAMPTZ NULL DEFAULT '2021-10-28 10:04:03.520917+00:00':::TIMESTAMPTZ,
	col30 BOOL NULL,
	col31 BOOL NULL,
	col32 BOOL NULL,
	col33 BOOL NULL DEFAULT false,
	col34 BOOL NULL,
	col35 BOOL NULL,
	col36 TIMESTAMPTZ NULL,
	col37 TIMESTAMPTZ NULL,
	col38 STRING NULL,
	col39 INT8 NULL,
	col40 STRING NULL,
	col41 UUID NULL,
	col42 INT8 NULL,
	col43 STRING NULL,
	col44 BOOL NULL,
	col45 BOOL NULL,
	CONSTRAINT "primary" PRIMARY KEY (id ASC),
	UNIQUE INDEX idx1 (col42 ASC),
	UNIQUE INDEX idx2 (col19 ASC) WHERE col19 != '':::STRING,
	INDEX idx3 (col26 ASC),
	UNIQUE INDEX idx4 (col3 ASC)
);
```
</details>

**Environment:**
 - CockroachDB version: 21.2.3
 - Server OS: Debian GNU/Linux 9.13
 - Client app: psql, python psycopg2, Golang pgx driver (all have the same behaviour)

**Comment by RaduBerinde** [comment]
@danchupin can you share the `ids.txt` file? I'm looking to double check that the IDs are unique.

**Comment by alxfv** [comment]
> @danchupin can you share the `ids.txt` file? I'm looking to double check that the IDs are unique.

Hi @RaduBerinde, I checked two times, IDs are unique.

**Comment by RaduBerinde** [comment]
@danchupin could you run a `SELECT COUNT(*) FROM my_table`? I just want to see how the real size compares with the one in our stats.

**Comment by RaduBerinde** [comment]
Also, if you look through the smaller result set, does it contain the smallest 28k ids, or do they seem more like a random subset of the 140k ids?

**Comment by RaduBerinde** [comment]
One more thing, if you play with increasing the batch size, is there a threshold after which it stops returning all the expected results?

Sorry for all the questions - given that we don't have everything we need to reproduce, I just want to make sure we have as much information as possible before we start working on reproducing it.

**Comment by alxfv** [comment]
> @danchupin could you run a `SELECT COUNT(*) FROM my_table`? I just want to see how the real size compares to the one in our stats.

There are 218.442.548 records.

> Also, if you look through the smaller result set, does it contain the smallest 28k ids, or do they seem more like a random subset of the 140k ids?

Seems random:

```
batch 0: intersection len with 28k result set = 1998
batch 1: intersection len with 28k result set = 1859
batch 2: intersection len with 28k result set = 1931
batch 3: intersection len with 28k result set = 1988
batch 4: intersection len with 28k result set = 1982
batch 5: intersection len with 28k result set = 2045
batch 6: intersection len with 28k result set = 1939
batch 7: intersection len with 28k result set = 2010
batch 8: intersection len with 28k result set = 2050
batch 9: intersection len with 28k result set = 1961
batch 10: intersection len with 28k result set = 2066
batch 11: intersection len with 28k result set = 1915
batch 12: intersection len with 28k result set = 1964
batch 13: intersection len with 28k result set = 2020
batch 14: intersection len with 28k result set = 321
```

**Comment by alxfv** [comment]
There are 532 ranges:

```
 select count(*) from [show ranges from table users];
 count
-------
   532
```

`select * from table where id in (...586k ids)` works as expected when all records are in the same range.

**Comment by RaduBerinde** [comment]
> Seems random:

That random distribution is expected if the `ids` are in arbitrary order - if you sort the ids, do the first batches intersect fully with the 28k result sets?

**Comment by alxfv** [comment]
> > Seems random:
> 
> That random distribution is expected if the `ids` are in arbitrary order - if you sort the ids, do the first batches intersect fully with the 28k result sets?

Yes:

```
batch 0: intersection len with 28k result set = 10000
batch 1: intersection len with 28k result set = 10000
batch 2: intersection len with 28k result set = 5888
batch 3: intersection len with 28k result set = 30
batch 4: intersection len with 28k result set = 27
batch 5: intersection len with 28k result set = 24
batch 6: intersection len with 28k result set = 36
batch 7: intersection len with 28k result set = 44
batch 8: intersection len with 28k result set = 28
batch 9: intersection len with 28k result set = 104
batch 10: intersection len with 28k result set = 501
batch 11: intersection len with 28k result set = 581
batch 12: intersection len with 28k result set = 347
batch 13: intersection len with 28k result set = 200
batch 14: intersection len with 28k result set = 239
```

**Comment by RaduBerinde** [comment]
Very interesting, thanks!

**Comment by danchupin** [comment]
[ids.txt](https://github.com/cockroachdb/cockroach/files/7860623/ids.txt)
@RaduBerinde here is the file.

**Comment by RaduBerinde** [comment]
I have been trying to reproduce this but have so far been unsuccessful. I will keep trying.

Let's try a few other things:
 - disable the vectorized execution engine: run `SET vectorize = off` before running the query (in the same session!); does it return all the results?
 - get a statement bundle, via `EXPLAN ANALYZE (DEBUG) SELECT id FROM my_table WHERE id IN (...)`. We can provide a way to upload the file for us (though you should check that any secret info is scrubbed, the file contains full schema info and traces can contain values from the table).

**Comment by alxfv** [comment]
Thank you very much for your time. 

- `SET vectorized = off;` didn't return all the results.
- do you need a specific file (like `trace.txt`) or the whole bundle?

**Comment by RaduBerinde** [comment]
The most important file is `trace-jaeger.json` but the others help make sure we're understanding what the query plan does exactly. The `distsql.html` diagram in particular is useful to understand the distribution of the plan.

**Comment by alxfv** [comment]
[obf-stmt-bundle-select-all-where-id-in-141k.zip](https://github.com/cockroachdb/cockroach/files/7894092/obf-stmt-bundle-select-all-where-id-in-141k.zip)

**Comment by alxfv** [comment]
Should be interesting: `select count(*) from users where user_id in (...140k ids)` returns 136k records. It's more than 28k, but still not all of them. 
[obf-stmt-bundle-select-count-where-id-in-141k.zip](https://github.com/cockroachdb/cockroach/files/7897101/obf-stmt-bundle-select-count-where-id-in-141k.zip)

**Comment by RaduBerinde** [comment]
Hi @AlexTalks - as KV L2, do you think you could take a look at the traces in the first bundle and see if anything stands out? This is a scan with a very large number of spans, executed on one node and it is returning only a subset of the expected KVs (about 28k instead of 140k). We know the KVs are there because the second query (which scans the same spans but is distributed differently) returns almost 140k results.

**Comment by tbg** [comment]
@RaduBerinde I'm not sure it's clear what Alex would be looking for, have you taken a look at the trace? Are we sure that both variations of the query are actually scanning the entire index? From the trace, can we see batches sent that cover the entire index?

**Comment by tbg** [comment]
also cc @abarganier, maybe something to learn here for the tracing teaching.

**Comment by tbg** [comment]
```
tobias@td:~/Downloads/obf-stmt-bundle-select-all-where-id-in-141k$ cat trace-jaeger.json | grep -E 'r[0-9]+: sending batch .*' | grep -Eo 'r[0-9]+:' | sort | uniq | wc -l
351

$ cat trace-jaeger.json | grep -E 'r[0-9]+: sending batch .*' | grep -Eo 'r[0-9]+:' | sort | uniq | wc -l
480
```

So the COUNT queries 480 ranges, the other one only 351.

Printing the rangeIDs for both and counting gives the difference. Everything with an 1 in front isn't queried by the `IN` flavor:

```
$ cat ~/Downloads/{count,where} | sort | uniq -c | sort -n | gh gist create
```

https://gist.github.com/bff4c1ea8795e8f4b61db367c8067ac6

**Comment by tbg** [comment]
`diff -u ~/Downloads/{count,where}` is also interesting. Unsurprisingly, it confirms that COUNT scans strictly more ranges (i.e. if IN visits, COUNT visits too). But the pattern is also interesting, by RangeID, it's interspersed; it's not like we're missing blocks of rangeIDs. Note that RangeIDs are assigned somewhat unpredictably, so we could still be missing one giant chunk of keyspace and the rangeIDs might not indicate that. Still, worth looking at.

https://gist.github.com/1858614e0f50e956d976a49c80cad39b

**Comment by tbg** [comment]
Just a hunch, but maybe the interesting ingredient here is that we're making batches that are mixing Scans and Gets:

> "value": "‹server/node.go:907 [n8] node received request: 311 Get, 1 Scan›"

I'll see if I can spot some patterns. For example, r9611 is hit only during the COUNT, with a batch of 115 Gets. r11135, 11 Gets. r3088, 308 Gets.

Now most batches are pure Gets:

```
tobias@td:~/Downloads/obf-stmt-bundle-select-count-where-id-in-141k$ cat trace-jaeger.json | grep 'sending batch .* to' | grep Scan | wc -l
134
tobias@td:~/Downloads/obf-stmt-bundle-select-count-where-id-in-141k$ cat trace-jaeger.json | grep 'sending batch .* to' | grep Get | wc -l
470
tobias@td:~/Downloads/obf-stmt-bundle-select-count-where-id-in-141k$ cat trace-jaeger.json | grep 'sending batch .* to' | grep Get | grep Scan | wc -l
122
```

But if we check every missing range, we find that each of them is pure gets:

```
tobias@td:~/Downloads/obf-stmt-bundle-select-count-where-id-in-141k$ for id in $(diff -u ~/Downloads/{count,where} | grep -E '^-.*' | sed s/-//g); do grep "${id} sending batch" trace-jaeger.json; done | grep Scan | wc -l
0
tobias@td:~/Downloads/obf-stmt-bundle-select-count-where-id-in-141k$ for id in $(diff -u ~/Downloads/{count,where} | grep -E '^-.*' | sed s/-//g); do grep "${id} sending batch" trace-jaeger.json; done | grep Get | wc -l
130
```

**Comment by tbg** [comment]
At this point it's going to get very manual. For example, we can pick this key from the read on r9611 in the COUNT, where we can find an explicit key that is being returned:

> "value": "‹kv/kvserver/replica_evaluate.go:521 [n3,s3,r9611/4:/Table/143/1/114{7096…-8112…}] evaluated Get command header:\u003ckey:\"\\366\\217\\211\\371\\006\\326\\214\\010\\210\" \u003e , txn=\"sql txn\" meta={id=c0bd840d key=/Min pri=0.02038167 epo=0 ts=1642571289.449805386,0 min=1642571289.449805386,0 seq=0} lock=false stat=PENDING rts=1642571289.449805386,0 wto=false gul=1642571289.949805386,0 : resp=header:\u003cnum_keys:1 num_bytes:413 \u003e value:\u003craw_bytes:..., err=\u003cnil\u003e›"

We see from that message that (KV) key `key:"\366\217\211\371\006\326\214\010\210"` returned a row (`num_keys=1`). Now we (someone) should be able to transcode that into the corresponding ID from the `IN` clause. We can then look through the `IN` trace to see which span should have queried that range, and perhaps in doing so develop a theory for why this isn't happening.

I wanted to get the ball rolling but will not be available to push it much further. If I had to guess, I would say everything is pointing to a range pagination bug in DistSender that is triggered by mixed Scan/Get batches. These have caused problems recently (@yuzefovich knows the relevant details) and perhaps there are issues where DistSender will skip over ranges. Or the problem could be in whatever sits on top of DistSender.

The `IN` trace is lighter on scans (per capita), maybe this has the bug trigger more often. Not a super compelling argument.

```
tobias@td:~/Downloads/obf-stmt-bundle-select-count-where-id-in-141k$ cat trace-jaeger.json | grep 'node received request' | grep Scan | wc -l
55
tobias@td:~/Downloads/obf-stmt-bundle-select-count-where-id-in-141k$ cat trace-jaeger.json | grep 'node received request' | grep Get | wc -l
338
tobias@td:~/Downloads/obf-stmt-bundle-select-count-where-id-in-141k$ cat trace-jaeger.json | grep 'node received request' | grep Scan | grep Get | wc -l
43
```

**Comment by tbg** [comment]
Sounds good! I think we have pretty good coverage for mixed get scan
batches in kvnemesis though come to think of it. Still, I'm not sure how
else to explain this, if the problem isn't in SQL.

(sent from mobile, please excuse my brevity)

**Comment by RaduBerinde** [comment]
Still haven't been able to repro so I'm spending some more time with the traces.

Looking at one of the missing ranges: r9611. In the COUNT traces:
```
querying next range at /Table/143/1/114710950/0›"
r9611: sending batch 115 Get to (n3,s3):4›"
```

We see this in the "all" traces:
```
329267:                 "value": "‹kv/kvclient/kvcoord/range_iter.go:175 [n1,client=10.0.0.126:57808,hostssl,user=root,txn=b054d471] querying next range at /Table/143/1/114408503›" 
329277:                 "value": "‹kv/kvclient/kvcoord/dist_sender.go:1916 [n1,client=10.0.0.126:57808,hostssl,user=root,txn=b054d471] r8100: sending batch 1 Scan to (n6,s6):3›" 
329297:                 "value": "‹kv/kvclient/kvcoord/range_iter.go:175 [n1,client=10.0.0.126:57808,hostssl,user=root,txn=b054d471] querying next range at /Table/143/1/116595210›"
329307:                 "value": "‹kv/kvclient/kvcoord/dist_sender.go:1916 [n1,client=10.0.0.126:57808,hostssl,user=root,txn=b054d471] r7644: sending batch 1 Scan to (n7,s7):3›"
```
So we're only asking for 1 Scan! This would correspond to the span `[/114408503 - /114408504]` in the query. And then we skip over a lot of input values up to 11659210, which is also a range in the query (`[/116595210 - /116595211]`).

**Comment by RaduBerinde** [comment]
This pattern starts here:
```
21973:                  "value": "‹kv/kvclient/kvcoord/dist_sender.go:1916 [n1,client=10.0.0.126:57808,hostssl,user=root,txn=b054d471] r9530: sending batch 476 Get, 1 Scan to (n7,s7):3›"
328757:                 "value": "‹kv/kvclient/kvcoord/range_iter.go:175 [n1,client=10.0.0.126:57808,hostssl,user=root,txn=b054d471] querying next range at /Table/143/1/90970692/0›"
328767:                 "value": "‹kv/kvclient/kvcoord/dist_sender.go:1916 [n1,client=10.0.0.126:57808,hostssl,user=root,txn=b054d471] r9530: sending batch 121 Get to (n7,s7):3›"
328787:                 "value": "‹kv/kvclient/kvcoord/range_iter.go:175 [n1,client=10.0.0.126:57808,hostssl,user=root,txn=b054d471] querying next range at /Table/143/1/91967067›"
328797:                 "value": "‹kv/kvclient/kvcoord/dist_sender.go:1916 [n1,client=10.0.0.126:57808,hostssl,user=root,txn=b054d471] r8026: sending batch 1 Scan to (n4,s4):3›"
328817:                 "value": "‹kv/kvclient/kvcoord/range_iter.go:175 [n1,client=10.0.0.126:57808,hostssl,user=root,txn=b054d471] querying next range at /Table/143/1/93790453›"
328827:                 "value": "‹kv/kvclient/kvcoord/dist_sender.go:1916 [n1,client=10.0.0.126:57808,hostssl,user=root,txn=b054d471] r3274: sending batch 1 Scan to (n8,s8):13›"
328847:                 "value": "‹kv/kvclient/kvcoord/range_iter.go:175 [n1,client=10.0.0.126:57808,hostssl,user=root,txn=b054d471] querying next range at /Table/143/1/97249712›"
328857:                 "value": "‹kv/kvclient/kvcoord/dist_sender.go:1916 [n1,client=10.0.0.126:57808,hostssl,user=root,txn=b054d471] r8690: sending batch 1 Scan to (n8,s8):7›"
```

**Comment by RaduBerinde** [comment]
Here we are sending 477 spans:

```
querying next range at /Table/143/1/90481154/0›"
r9530: sending batch 476 Get, 1 Scan to (n7,s7):3›"
```
The 477th value after 90481154 in the scan spans is 91123397. But the next request we see is:
```
querying next range at /Table/143/1/90970692/0›"
r9530: sending batch 121 Get to (n7,s7):3›"
```
which is inside the same range. So here we could not return all 477 results in a single response and we had to resume. 121 spans is indeed the number of spans in the query between 90970692 and 91123397. So far this all looks correct.

But the next request we see is skipping over a lot, and has only one scan! And everything that follows is similar.
```
querying next range at /Table/143/1/91967067›"
r8026: sending batch 1 Scan to (n4,s4):3›"
```

Indeed, this points to a bug in distsender or the kv fetcher. I don't know if there is much else to find out from the traces before trying to read through all the relevant code.

**Comment by RaduBerinde** [comment]
@yuzefovich I could use some help understanding the `kv_batch_fetcher.go` code. I sent some questions in #70608 - it's possible that PR actually is the culprit but I don't know.

For the record, this is the list of changes to that file between v21.1.7 and v21.2.3:
6f88d4f6b8 row: improve the txnKVFetcher
e45862cd43 row,rowinfra: refactor some types and constants out of row
6572d21fc7 sql: remove unused field in txnKVFetcher
0fa923a65c sql: rationalize batch key and bytes limit
a017cc7e6a sql: rename utility function in fetcher
edf5dafc98 kvfollowerreadsccl: plug bounded staleness into the execution engine
38da713040 sql: support lock_timeout session variable
a5708ab8da kv,storage: use a BytesMonitor to track memory allocations for scans
68beda64d8 sql,server: add admission control for DistSQL responses
8f65a257f7 Merge #66376
960817027c rowexec: add comments around joinReaderStrategies
8eba7b9f18 colfetcher: don't hold batches across boundaries
ba420215cf row: release memory more eagerly in txnKVFetcher
d6d394bf5c row: release memory more eagerly in txnKVFetcher
4e4f31a0ed *: integrate admission control for SQL=>KV and KV=>SQL
5b7a6102a1 sql: issue GetRequests from SQL when possible

**Comment by RaduBerinde** [comment]
I believe that the problem is that the kv batch fetcher assumes that the `ResumeSpan` will be set for all requests that were not satisfied. But dist sender only fills in the ResumeSpan for scans: https://github.com/cockroachdb/cockroach/blob/cb0222a1980b96ad59c181c96fab8340aa252571/pkg/kv/kvclient/kvcoord/dist_sender.go#L1712

This explains why we only see Scans after the first `fetch()`.

**Comment by RaduBerinde** [comment]
Probably introduced by #61583 when we started using `Get`s, although it's possible some of the subsequent cleanup exposed it.

**Comment by yuzefovich** [comment]
@danchupin @alxfv thanks for your cooperation for getting to the bottom of this issue! I think Radu (with a little bit help from me) narrowed [down the bug](https://github.com/cockroachdb/cockroach/pull/70608#pullrequestreview-860126499), and I believe we have a repro and understand the circumstances when the bug occurs.

We'll issue a fix shortly, but in the meantime a possible workaround it to limit the number of `id`s in the `IN` clause to not exceed 10k.

**Comment by RaduBerinde** [comment]
Here is a simple repro (with `cockroach demo --nodes=2`):

```
CREATE TABLE t (k INT PRIMARY KEY, blob STRING); 
ALTER TABLE t SPLIT AT VALUES (1000000);        
ALTER TABLE t EXPERIMENTAL_RELOCATE VALUES (ARRAY[1], 0), (ARRAY[2], 1000000);
INSERT INTO t SELECT x, repeat('a', 100) FROM generate_series(1, 2000000) AS g(x);
SHOW RANGES FROM TABLE t;
SELECT COUNT(*) FROM t WHERE (k >= 600000 AND k < 1200000) OR k=1500000;
SET DISTSQL = OFF;
SELECT COUNT(*) FROM t WHERE (k >= 600000 AND k < 1200000) OR k=1500000;
```

Output:
```
  start_key | end_key  | range_id | range_size_mb | lease_holder | lease_holder_locality | replicas |    replica_localities
------------+----------+----------+---------------+--------------+-----------------------+----------+---------------------------
  NULL      | /1000000 |       53 |    206.281892 |            1 | region=us-east1,az=b  | {1}      | {"region=us-east1,az=b"}
  /1000000  | NULL     |       54 |    203.793209 |            2 | region=us-east1,az=c  | {2}      | {"region=us-east1,az=c"}
(2 rows)

  count
----------
  600001
(1 row)

  count
----------
  600000
(1 row)
```

In the last case, we drop the Get for `k=1500000`.

**Comment by alxfv** [comment]
Thanks everyone for dedicating the time to this!

**Comment by RaduBerinde** [comment]
This will be fixed in 21.2.5, which currently is planned for Feb 7.

**Comment by danchupin** [comment]
@RaduBerinde thank you!
